arangod/Pregel/Actor/include/Actor/Message.h

```diff
@@ -0,0 +1,134 @@
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Volmer
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include <Inspection/VPackWithErrorT.h>
+#include <memory>
+#include <variant>
+
+#include <velocypack/Builder.h>
+
+#include "ActorPID.h"
+#include "MPSCQueue.h"
+
+namespace {
+// helper type for the visitor
+template<class... Ts>
+struct overloaded : Ts... {
+  using Ts::operator()...;
+};
+template<class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+}  // namespace
+
+namespace arangodb::pregel::actor {
+
+struct MessagePayloadBase {
+  virtual ~MessagePayloadBase() = default;
+};
+
+template<typename Payload>
+struct MessagePayload : MessagePayloadBase {
+  template<typename... Args>
+  MessagePayload(Args&&... args) : payload(std::forward<Args>(args)...) {}
+
+  Payload payload;
+};
+template<typename Payload, typename Inspector>
+auto inspect(Inspector& f, MessagePayload<Payload>& x) {
+  return f.object(x).fields(f.field("payload", x.payload));
+}
+
+struct UnknownMessage {
+  ActorPID sender;
+  ActorPID receiver;
+};
+template<typename Inspector>
+auto inspect(Inspector& f, UnknownMessage& x) {
+  return f.object(x).fields(f.field("sender", x.sender),
+                            f.field("receiver", x.receiver));
+}
+
+struct ActorNotFound {
+  ActorPID actor;
+};
+template<typename Inspector>
+auto inspect(Inspector& f, ActorNotFound& x) {
+  return f.object(x).fields(f.field("actor", x.actor));
+}
+
+struct ServerNotFound {
+  ServerID server;
+};
+template<typename Inspector>
+auto inspect(Inspector& f, ServerNotFound& x) {
+  return f.object(x).fields(f.field("server", x.server));
+}
+
+struct ActorError
+    : std::variant<UnknownMessage, ActorNotFound, ServerNotFound> {
+  using std::variant<UnknownMessage, ActorNotFound, ServerNotFound>::variant;
+};
+template<typename Inspector>
+auto inspect(Inspector& f, ActorError& x) {
+  return f.variant(x).unqualified().alternatives(
+      arangodb::inspection::type<UnknownMessage>("UnknownMessage"),
+      arangodb::inspection::type<ActorNotFound>("ActorNotFound"),
+      arangodb::inspection::type<ServerNotFound>("ServerNotFound"));
+}
+
+template<typename T, typename U>
+struct concatenator;
+template<typename... Args0, typename... Args1>
+struct concatenator<std::variant<Args0...>, std::variant<Args1...>> {
+  std::variant<Args0..., Args1...> item;
+  concatenator(std::variant<Args0...> a) {
+    std::visit(overloaded{[this](auto&& arg) { this->item = std::move(arg); }},
+               std::move(a));
+  }
+  concatenator(std::variant<Args1...> b) {
+    std::visit(overloaded{[this](auto&& arg) { this->item = std::move(arg); }},
+               std::move(b));
```

Comment:
I'm pretty sure overloaded don't needed here, please excuse me and provide compile error if I'm wrong.

```suggestion
    std::visit([this](auto&& arg) { this->item = std::move(arg); },
               std::move(a));
  }
  concatenator(std::variant<Args1...> b) {
    std::visit([this](auto&& arg) { this->item = std::move(arg); },
               std::move(b));
```